CP-50914 and CP-50717: Multi-version Drivers: Toolstack Implementation and extend bugtool data collection

last-genius opened the pull request:
Implements the feature outlined in the toolstack design document.

* New class `Host_driver`, with `select` and `deselect` methods
* New `Host` method `rescan-drivers`

Contains temporary workarounds intended to be removed when the foundations work is done.

last-genius commented:
Not sure if it's worth creating .mli files for the failed test or just incrementing the number in the quality gate

psafont commented:
It's very easy to create empty .mli files :)

last-genius commented:
Yeah, of course, but not sure if it's worth populating an already busy directory with empty files.

lindig commented:
We should introduce MLI files; empty or not.

lindig commented:
I would suggest to keep the schema hash update in a single commit - this most likely will need updating for a merge into master.

last-genius commented:
having an empty .mli file in ocaml/idl breaks everything

lindig commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,276 @@
+  in
+  let driver_versions =
+    let arr = Sys.readdir (all_versions_dir // driver_name) in
+    Array.sort __compare_versions arr ;
```
Could also use `Array.to_list |> List.sort`

lindig commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,276 @@
+  in
+  (selected_version, active_version, driver_versions)
+
+let discover ~__context ~host =
```
Needs a comment what this does

last-genius commented on ocaml/xapi/xapi_host_driver_helpers.ml:
```diff
@@ -0,0 +1,133 @@
+  n + (-n & 3)
+
+(** advance the cursor to position n *)
+let advance_to n =
+  let* p = pos in
+  advance (max 0 (n - p))
+
+(** align the cursor to a multiple of 4 *)
+let align =
+  let* p = pos in
+  advance_to (align4 p)
```
Interestingly, our formatter turns `let* pos in` into `let* pos = pos in`....

last-genius commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,276 @@
+type driver_paths = {selected_dir: string; all_versions_dir: string}
+
+let __get_drivers_base_path ~__context ~host =
+  let linux_version =
```
This is what xapi does to obtain this DB row too

lindig commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,276 @@
+      in
+      let destfile = driver_paths.selected_dir // (driver_name ^ ".ko") in
+      try
+        if Sys.file_exists destfile then
```
It does not matter here, but this created a race between test and action. It's better to always remove and to catch the failure.

last-genius commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,276 @@
+  }
+
+let __reload_driver_selection ~__context ~host =
+  let linux_version =
```
weirdly enough, trying to do this when the module loads causes these obscure errors:
```ocaml
File "ocaml/tests/dune", line 85, characters 56-80:
85 |     test_xapi_helpers test_tar_ext test_pool_repository test_host_driver_helpers)
                                                             ^^^^^^^^^^^^^^^^^^^^^^^^
Fatal error: exception Sys_error("/etc/28fdea.tmp: Permission denied")
File "ocaml/tests/dune", line 157, characters 6-19:
157 | (name test_observer)
            ^^^^^^^^^^^^^
Fatal error: exception Sys_error("/etc/c83e4a.tmp: Permission denied")
```
This is how I obtained the constant:
```ocaml
let linux_version =
  let __context = Context.make "xapi_host_driver:linux_version" in
  let host = Helpers.get_localhost ~__context in
  List.assoc "linux" (Db.Host.get_software_version ~__context ~self:host)
```
Reverting this change and obtaining linux_version inside functions with real context fixes these errors.

last-genius commented:
Addressed most review comments in fixup commits, best reviewed individually. These will be squashed before merging.

Added a CLI binary that lists parsed ELF notes for multi-version drivers, this is included in the bugtool's output.

Will test a bit more tomorrow.

The only thing left is to improve the handling of versions - split them at the dots, etc. - will address this separately.

last-genius commented:
Not sure why SDK CI is failing. DCO will be fixed when I squash the commits, and formatting tests would require me to merge master into the feature branch.

Vincent-lau commented on ocaml/xapi/message_forwarding.ml:
```diff
@@ -6554,6 +6554,40 @@ functor
+      let create ~__context ~host ~name_label ~versions ~active_version
+          ~selected_version =
+        info "%s" __FUNCTION__ ;
+        Xapi_host_driver.create ~__context ~host ~name_label ~versions
```
Is it possible that a pool member receives this call? Do you need to have some kind of helper to always forward this call to the coordinator?

lindig replied:
I believe an API call is only accepted on the coordinator - but we can ask @robhoes if the forwarding mechanism is required.

robhoes replied:
The `create` and `destroy` functions are not actually part of the API as defined in the datamodel, as `~gen_constructor_destructor:false` and the are not explicitly defined. And we don't need them, as xapi itself is meant to create and destroy these objects. So these functions (in message_forwarding.ml) will never be executed.

Vincent-lau commented on ocaml/idl/schematest.ml:
```diff
@@ -3,7 +3,7 @@ let hash x = Digest.string x |> Digest.to_hex
    ocaml/idl/datamodel_common.ml, usually schema_minor_vsn *)
 
-let last_known_schema_hash = "428caff23cdb969c59a9960beefd7bb6"
+let last_known_schema_hash = "88ce88a5c467ddfb8e65d9bdacd3d67a"
```
Not a big deal, but I suggest updating this as part of the commit that changes the datamodel, so we get unit tests pass on all commits

last-genius replied:
I was previously asked to put this into a separate commit (it was part of the datamodel commit initially) because it'd need changing closer to merge into master.

Vincent-lau commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -249,12 +249,13 @@ let discover ~__context ~host =
   | Sys_error e ->
-      internal_error "Failed to parse drivers on host %s: %s"
-        (Ref.string_of host) e
+      D.debug "Failed to parse drivers on host %s: %s" (Ref.string_of host) e ;
```
Maybe add a TODO here so that we know this needs to be changed once the foundation work is done?

last-genius replied:
It's in a separate temporary commit that will just be dropped.

Vincent-lau commented on ocaml/idl/datamodel_host.ml:
```diff
@@ -1766,6 +1766,13 @@ let apply_updates =
     ~allowed_roles:(_R_POOL_OP ++ _R_CLIENT_CERT)
     ()
 
+let rescan_drivers =
```
so the reason to put `rescan_driver` in the host class is that it can only be called on hosts. How about `select` and `deselect`, can they only be called on hosts?

last-genius replied:
The driver uuid already encodes which host it's located on; specifying both the host uuid and the driver uuid would be redundant - the call already has the information necessary to determine where to forward it.

Vincent-lau commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -85,6 +94,168 @@ let __reload_driver_selection ~__context ~host =
+      (split a) (split b)
+end
+
+let __get_driver_versions driver_name selected_dir all_versions_dir =
```
why `__`?

last-genius replied:
Just a Python-style indication of an internal helper function given that `.mli` files do not play nice in this directory, I believe.

Vincent-lau replied:
ok, but I have never seen this kind of style being used on ocaml functions, maybe @lindig would know. What is wrong with the mli files?

lindig replied:
I've not seen the `_` prefixing - it suppresses the `unused function` warning, so would not use it to designate internal functions. But I'm a live-and-let-live guy.

last-genius replied:
I've removed the underscores and created an `.mli` file

robhoes commented on ocaml/idl/datamodel_host_driver.ml:
```diff
@@ -0,0 +1,67 @@
+        uid _host_driver
+      ; field ~qualifier:StaticRO ~ty:(Ref _host) "host"
+          "Host where this driver is installed"
+      ; field ~qualifier:DynamicRO ~ty:String "name_label"
```
We use `name_label` along with `name_description` fields for user-defined text identifiers, as read/write fields. But this is different, as it is the actual name of a driver coming from the OS below. You could call it just `name` to avoid confusion (as the design has).

last-genius replied:
As far as I remember, the reason for renaming it into name-label was to get nice autocompletion by default (it knows how to do it for name-label, but not for name since almost nothing else uses it). Will change it back!

robhoes commented on ocaml/idl/datamodel_host_driver.ml:
```diff
@@ -0,0 +1,67 @@
+    ~contents:
+      [
+        uid _host_driver
+      ; field ~qualifier:StaticRO ~ty:(Ref _host) "host"
```
Please add `~lifecycle:[]` for all fields.

robhoes commented on ocaml/xapi-cli-server/cli_operations.ml:
```diff
@@ -1322,6 +1322,18 @@ let gen_cmds rpc session_id =
+          ; "versions"
+          ; "active-version"
+          ; "selected-version"
+          ; "requires-reboot-to-become-active"
```
I prefer the shorter version too.

robhoes commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,140 @@
+
+(* Runs on the necessary host *)
+let select ~__context ~self ~version =
+  let versions = Db.Host_driver.get_versions ~__context ~self in
```
This is part of the first commit, which says "empty implementation", but it definitely isn't empty ;)

robhoes commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,140 @@
+  in
+  try
+    let _, _ =
+      Forkhelpers.execute_command_get_output "/usr/sbin/depmod"
```
We should define these command names and paths in in xapi_globs, using the resources mechanism.

robhoes commented on ocaml/xapi-cli-server/cli_frontend.ml:
```diff
@@ -1045,6 +1045,15 @@ let rec cmdtable_data : (string * cmd_spec) list =
+      ; optn= []
+      ; help= "Scan the host and update its driver information."
+      ; implementation= No_fd Cli_operations.host_rescan_drivers
+      ; flags= []
```
You can use `Host_selectors` here rather than the `uuid` arg above, since this is on the host. It's a bit more flexible for users.

robhoes commented on ocaml/idl/datamodel_host.ml:
```diff
@@ -1766,6 +1766,13 @@ let apply_updates =
+  call ~name:"rescan_drivers" ~in_oss_since:None ~lifecycle:[]
+    ~doc:"Scan the host and update its driver information."
+    ~params:[(Ref _host, "host", "The host to be rescanned")]
+    ~result:(Map (String, Set String), "Updates to the drivers' versions")
```
The return type is not obvious. The implementation defines some "keys" for what is returned, but the API documentation won't show that. I wonder how useful it is to return this data anyway; I had expected no return value. Especially since these rescans also happen implicitly on restart, where you'd never see any of it.

last-genius replied:
I can specify the return keys in the API docs. I find it a pretty user-friendly addition, and a return value can always be ignored if the user isn't interested in it.

robhoes commented on ocaml/xapi/xapi_host.ml:
```diff
@@ -3027,6 +3027,8 @@ let apply_updates ~__context ~self ~hash =
   Db.Host.set_last_update_hash ~__context ~self ~value:hash ;
   warnings
 
+let rescan_drivers ~__context ~host = Xapi_host_driver.discover ~__context ~host
```
Or even just `let rescan_drivers = Xapi_host_driver.discover`.

robhoes commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -85,6 +87,160 @@ let reload_driver_selection ~__context ~host =
+*)
+module Version = struct
+  let compare a b =
+    let split str = Str.(split_delim (regexp "[.,:;+_-]") str) in
```
Is `Str` (thread) safe to use? I can never remember, but always get worried when I see it... :/

last-genius replied:
I think this particular function should be thread-safe, but @edwintorok would know better. Have the issues with `Re` submodules we thought were thread-safe but ended up not being thread-safe been resolved?

robhoes commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -85,6 +87,160 @@ let reload_driver_selection ~__context ~host =
+  try
+    let driver_paths = get_drivers_base_path ~__context ~host in
+    let db_drivers_set, db_drivers_map =
+      Db.Host_driver.get_all_records ~__context
```
You could use `get_all_records_where` matching on the `host` (and drop the if-statement below) to restrict the amount of data being transferred from the coordinator.

last-genius replied:
I couldn't really grok `get_all_records_where` so decided to use this version, turns out this is an old problem: https://lists.xenproject.org/archives/html/xen-api/2013-02/msg00068.html

robhoes replied:
Yes, it's odd, and best to look for examples in the code base.

robhoes commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -85,6 +87,160 @@ let reload_driver_selection ~__context ~host =
+                 driver_paths.all_versions_dir
+             in
+             let selected_changed =
+               Db.Host_driver.get_selected_version ~__context ~self
```
This and the DB-get calls below trigger potentially very many round-trips to the coordinator. But this seems unnecessary, as all the required data has been fetched from the DB already using `get_all_records` above?

robhoes commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,297 @@
+         )
+    in
+    [
+      ("updated_drivers", updated_drivers)
+    ; ("removed_drivers", removed_drivers)
+    ; ("added_drivers", added_drivers)
```
I think it is better to log these changes rather than return them to the API caller, which in the common case would be internal (dbsync).

minglumlu replied:
Yeah. It seems to me that the data in xapi db is a read-only cache of the situation on the filesystem. E.g. even there is inconsistency between xapi db and filesystem, the xapi db will be updated according to the filesystem always. 
So would a much simpler approach be just to flush the cache completely and refill it, given there will not be too many such drivers?

lindig replied:
> So would a much simpler approach be just to flush the cache completely and refill it, given there will not be too many such drivers?

If we flush the cache and re-create it, we also invalidate all references to drivers that an API client could hold. I would like to avoid that.

stormi commented:
> Implements the feature outlined in the toolstack design document.

Is this document public?

minglumlu commented on ocaml/idl/datamodel_host_driver.ml:
```diff
@@ -0,0 +1,67 @@
+    ~contents:
+      [
+        uid _host_driver
+      ; field ~qualifier:StaticRO ~ty:(Ref _host) "host"
```
Will this r`ef host` make sure that all the `Host_driver` objects of this host be removed when the host is ejected from a pool?

robhoes replied:
Good point: they won't get cleaned up, unless we add a rule to `Db_gc`.
We could add a `(host_driver ref set) host.drivers` field and a relation to the datamodel to make this easier.

minglumlu commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,297 @@
+      []
+
+(* Runs on the necessary host *)
+let select ~__context ~self ~version =
```
Would the `select`/`deselect` need a protection from being invoked concurrently?

minglumlu replied:
Hmm. I think the symbol link under `/lib/modules/<kernel-version>/updates/` needs protection as well.

last-genius replied:
This is somewhat harder than a single mutex per reload, since we'd have to protect a particular driver name - a different driver could still be changed concurrently, right? I do not think anything particularly harmful can come from concurrent removal/symlinking of the same driver - the last reload will pick up the last changes anyway.

minglumlu replied:
> the last reload will pick up the last changes anyway.

I see. Thanks.

minglumlu commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,297 @@
+             let versions_changed =
+               Db.Host_driver.get_versions ~__context ~self <> driver_versions
+             in
+             if selected_changed then
```
Would the `reload_driver_selection` need to be called for this `selected_changed`? Or need to find out the version in the `initrd` for thorough consistency?

last-genius replied:
My impression was that rescan is primarily for discovering installer-created file hierarchy, with depmod and the like already called. We do not really (want to) support changes to multi-version drivers behind xapi's back, so we assume if files were moved around without proper changes to the DB (not through `select`, `deselect` calls), depmod and the like would also be called. Adding logic to parse initrd too for an officially unsupported use-case might be too much.

minglumlu commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,297 @@
+        reload_driver_selection ~__context ~host ;
+
+        if Db.Host_driver.get_active_version ~__context ~self = "" then
+          Db.Host_driver.set_active_version ~__context ~self ~value:version ;
```
Would the driver have been active? I understand it would be only after the host rebooted.

last-genius replied:
Thank you for noticing this! After talking to Foundations - if the driver is required by the hardware, but none have been active, a selected version can immediately become active without reboot. But we do need to run a command to re-trigger the scan for required drivers and actually check if the driver has been loaded - I will add this.

minglumlu commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,297 @@
+  in
+  try
+    let _, _ =
+      Forkhelpers.execute_command_get_output "/usr/sbin/depmod"
```
I'm not sure if this `depmod` could load the driver module into the running Kernel or not.
Do we need to reboot the host to get the module loaded into the kernel actually, or use `modprobe`?

last-genius replied:
We don't necessarily need to reboot, you're right. `modprobe` operates on a single driver, we will need to trigger a udev event for device drivers to be added again instead.

minglumlu replied:
Thanks. Now I refresh my mind as the following. Please correct me if I'm wrong.
1. The device is already equipped on the host, when doing `select`,  the symbol link under `/lib/modules/<kernel-version>/updates/` is (re)created, and the udevadm command can trigger the Kernel to reload the new driver for the device.
2. The device is not equipped on the host yet, when doing `select`, the symbol link is (re)created, but no need to load it since it will be automatically when the device is present. But udevadm is harmless in this case.

last-genius replied:
Precisely!

last-genius commented:
Addressed most of the review comments in new and fixup commits. Aside from stylistic changes and renaming, larger changes are:
* Added Host_driver cleanup to `db_gc`
* Added a rescan command that is needed to load required but previously absent drivers
* This requires verifying if a driver has been activated after selection of if it's necessary to wait for a reboot, therefore split `discover` into two functions, with the helper one called from `select` and `discover`.
* Added a mutex for the `reload_driver_selection` function to prevent possible conflicts in dependency and initrd generation.
* Added the design document into the docs in a separate commit @stormi 

Remaining review comments:
* Whether to return information from `discover` or just log it
* Is `Str.split_delim` thread-safe?

Will test to see if I haven't introduced any regressions.

stormi commented on doc/content/toolstack/features/MVD/index.md:
```diff
@@ -0,0 +1,292 @@
+
+* `/sys/module/xfs/notes/.note.gnu.build-id`
+
+with a specific name (`.note.xenserver`) for our purpose. Such a file contains
+in binary encoding a sequence of records, each containing:
```
There may be political implications to my remark, but I think every XenServer mention in this PR, including in the `/lib/modules/<kernel-version>/xenserver/<driver>/<version>/<name>.ko` path and in the ELF notes, could be replaced by `XAPI` or `xapi`, and I think this would be more accurate since this is a XAPI mechanism.

stormi commented:
This looks like an interesting feature. At XCP-ng, we already provide alternate versions of various drivers, but the design is nothing like yours here. We simply have optional RPMs in our repositories, that users can install, and these add themselves in `/lib/modules/<kernel_version>/override`, but it's less flexible, and XAPI integration will likely be a real plus.

Will it also offer the ability to disabled a vendor driver in order to fall back to the built-in one? With some vendor with buggy drivers or no intent to support older hardware, this sometimes proves useful.

minglumlu commented on ocaml/xapi/xapi_host_driver.ml:
```diff
@@ -0,0 +1,338 @@
+      DriverSet.inter db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.filter_map (fun driver_name ->
+             let self, db_driver = DriverMap.find driver_name db_drivers_map in
```
If `DriverMap.find` is used here anyway, then the `DriverSet.inter` seems not necessary. The logic would be: for each driver in `filesystem_drivers`, `DriverMap.find_opt driver_name db_drivers_map |> Option.map (fun (ref, record) -> ...)`.

minglumlu commented:
In a RPU case, it's not clear for me on how to ensure the upgraded XenServer keeping using the selected version?

lindig commented:
I think we should merge this and address any remaining issues in a new PR

last-genius commented:
Squashed the fixup commits, will address remaining comments separately.
